memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala

```diff
@@ -0,0 +1,295 @@
+package filodb.memory.format.vectors
+
+import org.agrona.concurrent.UnsafeBuffer
+import scalaxy.loops._
+
+import filodb.memory.BinaryRegion
+import filodb.memory.format._
+
+/**
+ * A trait to represent bucket-based histograms as well as derived statistics such as sums or rates of
+ * increasing histograms.
+ * The schema is based on Prometheus histograms.  Each bucket is designed to contain all observations less than
+ * or equal to the bucketTop ("upper bound"), thus it is cumulative with increasing bucket number.
+ * Furthermore the LAST bucket should contain the total number of observations overall.
+ */
+trait Histogram extends Ordered[Histogram] {
+  def numBuckets: Int
+
+  /**
+   * Gets the bucket definition for number no.  Observations for values <= this bucket, so it represents
+   * an upper limit.
+   */
+  def bucketTop(no: Int): Double
+
+  /**
+   * Gets the counter or number of observations for a particular bucket, or derived value of a bucket.
+   */
+  def bucketValue(no: Int): Double
```

Comment:
It may be either.  Depends on the input type.  For this reason it cannot be made into "rate" vs "bucketValue" because it depends on the calculation.    Maybe we should discuss offline.